packages/core/src/routes/interaction/utils/sign-in-experience-validation.ts

```diff
@@ -0,0 +1,131 @@
+import type { Event, SignInExperience, Profile } from '@logto/schemas';
+import { SignUpIdentifier, SignInMode, SignInIdentifier } from '@logto/schemas';
+
+import RequestError from '#src/errors/RequestError/index.js';
+import assertThat from '#src/utils/assert-that.js';
+
+import type { IdentifierPayload } from '../types/guard.js';
+
+const forbiddenEventError = new RequestError({ code: 'auth.forbidden', status: 403 });
+
+const forbiddenIdentifierError = new RequestError({
+  code: 'user.sign_in_method_not_enabled',
+  status: 422,
+});
+
+export const signInModeValidation = (event: Event, { signInMode }: SignInExperience) => {
+  if (event === 'sign-in') {
+    assertThat(signInMode !== SignInMode.Register, forbiddenEventError);
+  }
+
+  if (event === 'register') {
+    assertThat(signInMode !== SignInMode.SignIn, forbiddenEventError);
```

Comment:
nit: Why the `event` type is a string union, whereas the `signInMode` is an enum?

Reply:
good point, some legacy issue. The original signInMode is to distinguish the allowed method of sign-in and sign-up. But the event is more like for the client to specify the request event sign-in, sign-up, or forgot password. So strictly saying they are not sharing the same definition scope. 

Also, for the new event type, using string union could avoid the dependency of an enum object. We could import a type instead of importing an enum object.

Reply:
Well, then leave it as-is